### PR TITLE
Convert shadow and filter types to strongly typed CSS values

### PR DESCRIPTION
**Pull request by weinig**
#### b5626802cc009d8ecc6d16402759c2aa0745eb59
<pre>
Convert shadow and filter types to strongly typed CSS values
<a href="https://bugs.webkit.org/show_bug.cgi?id=283921">https://bugs.webkit.org/show_bug.cgi?id=283921</a>

Reviewed by NOBODY

Converts the shadow and filter types to strongly typed CSS values.

Formerly, CSS shadows for `box-shadow` and `text-shadow` use the
same representation for CSS, a CSSValueList of CSSShadowValues.
Now, they have been split, to make it clear that `text-shadow`
supports a different set of functionality than `box-shadow`.

Similarly, CSS filters for both `filter` and `-apple-color-filter`
use the same representation, a CSSValueList of CSSFunctionValues.
Now they too have been split to make it clear that the set of
filter functions supported is different for the different properties.

Unlike previous conversions, this one focuses primarily on the
CSS side, as the representations used for shadows and filters
in style system are a bit tangled. However, structure on the
style side is put in place, and some modest wins are present,
such as the shrinking of the ShadowData type due to use of the new
Style::Length types.

To maintain backwards compatibility, extra plumbing was required
to construct DeprecatedCSSOMValues matching the shipping types.

* Source/WTF/wtf/StdLibExtras.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/SaferCPPExpectations/RefCntblBaseVirtualDtorExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
* Source/WebCore/css/CSSAppleColorFilterPropertyValue.cpp: Added.
* Source/WebCore/css/CSSAppleColorFilterPropertyValue.h: Added.
* Source/WebCore/css/CSSBoxShadowPropertyValue.cpp: Added.
* Source/WebCore/css/CSSBoxShadowPropertyValue.h: Added.
* Source/WebCore/css/CSSFilterImageValue.cpp:
* Source/WebCore/css/CSSFilterImageValue.h:
* Source/WebCore/css/CSSFilterPropertyValue.cpp: Added.
* Source/WebCore/css/CSSFilterPropertyValue.h: Added.
* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSShadowValue.cpp: Removed.
* Source/WebCore/css/CSSShadowValue.h: Removed.
* Source/WebCore/css/CSSTextShadowPropertyValue.cpp: Added.
* Source/WebCore/css/CSSTextShadowPropertyValue.h: Added.
* Source/WebCore/css/CSSValue.cpp:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/css/CSSValuePool.cpp:
* Source/WebCore/css/CSSValuePool.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/css/ComputedStyleExtractor.h:
* Source/WebCore/css/DeprecatedCSSOMBoxShadowValue.cpp: Added.
* Source/WebCore/css/DeprecatedCSSOMBoxShadowValue.h: Added.
* Source/WebCore/css/DeprecatedCSSOMFilterFunctionValue.cpp: Added.
* Source/WebCore/css/DeprecatedCSSOMFilterFunctionValue.h: Added.
* Source/WebCore/css/DeprecatedCSSOMTextShadowValue.cpp: Added.
* Source/WebCore/css/DeprecatedCSSOMTextShadowValue.h: Added.
* Source/WebCore/css/DeprecatedCSSOMValue.h:
* Source/WebCore/css/DeprecatedCSSOMValueList.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Background.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Filter.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Filter.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+MetaConsumerDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TextDecoration.cpp:
* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/css/values/backgrounds/CSSBoxShadow.h: Added.
* Source/WebCore/css/values/backgrounds/CSSBoxShadowProperty.h: Added.
* Source/WebCore/css/values/color/CSSColor.h:
* Source/WebCore/css/values/filter-effects/CSSAppleColorFilterProperty.h: Added.
* Source/WebCore/css/values/filter-effects/CSSAppleInvertLightnessFunction.h: Added.
* Source/WebCore/css/values/filter-effects/CSSBlurFunction.h: Added.
* Source/WebCore/css/values/filter-effects/CSSBrightnessFunction.h: Added.
* Source/WebCore/css/values/filter-effects/CSSContrastFunction.h: Added.
* Source/WebCore/css/values/filter-effects/CSSDropShadowFunction.h: Added.
* Source/WebCore/css/values/filter-effects/CSSFilterFunction.h: Added.
* Source/WebCore/css/values/filter-effects/CSSFilterFunctionDescriptor.h: Renamed from Source/WebCore/css/CSSFilterFunctionDescriptor.h.
* Source/WebCore/css/values/filter-effects/CSSFilterProperty.h: Added.
* Source/WebCore/css/values/filter-effects/CSSFilterReference.cpp: Added.
* Source/WebCore/css/values/filter-effects/CSSFilterReference.h: Added.
* Source/WebCore/css/values/filter-effects/CSSGrayscaleFunction.h: Added.
* Source/WebCore/css/values/filter-effects/CSSHueRotateFunction.h: Added.
* Source/WebCore/css/values/filter-effects/CSSInvertFunction.h: Added.
* Source/WebCore/css/values/filter-effects/CSSOpacityFunction.h: Added.
* Source/WebCore/css/values/filter-effects/CSSSaturateFunction.h: Added.
* Source/WebCore/css/values/filter-effects/CSSSepiaFunction.h: Added.
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericRange.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+CSSValueVisitation.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+ComputedStyleDependencies.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Serialization.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes.h:
* Source/WebCore/css/values/text-decoration/CSSTextShadow.h: Added.
* Source/WebCore/css/values/text-decoration/CSSTextShadowProperty.h: Added.
* Source/WebCore/editing/Editor.cpp:
* Source/WebCore/editing/FontAttributeChanges.cpp:
* Source/WebCore/html/canvas/CanvasRenderingContext2D.cpp:
* Source/WebCore/rendering/BackgroundPainter.cpp:
* Source/WebCore/rendering/BackgroundPainter.h:
* Source/WebCore/rendering/EllipsisBoxPainter.cpp:
* Source/WebCore/rendering/TextDecorationPainter.cpp:
* Source/WebCore/rendering/TextPainter.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/ShadowData.cpp:
* Source/WebCore/rendering/style/ShadowData.h:
* Source/WebCore/rendering/style/StyleFilterImage.cpp:
* Source/WebCore/style/FilterOperationsBuilder.cpp: Removed.
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleBuilderState.cpp:
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/backgrounds/StyleBoxShadow.cpp: Added.
* Source/WebCore/style/values/backgrounds/StyleBoxShadow.h: Added.
* Source/WebCore/style/values/color/StyleColor.h:
* Source/WebCore/style/values/filter-effects/StyleAppleColorFilterProperty.cpp: Added.
* Source/WebCore/style/values/filter-effects/StyleAppleColorFilterProperty.h: Added.
* Source/WebCore/style/values/filter-effects/StyleAppleInvertLightnessFunction.cpp: Added.
* Source/WebCore/style/values/filter-effects/StyleAppleInvertLightnessFunction.h: Added.
* Source/WebCore/style/values/filter-effects/StyleBlurFunction.cpp: Added.
* Source/WebCore/style/values/filter-effects/StyleBlurFunction.h: Added.
* Source/WebCore/style/values/filter-effects/StyleBrightnessFunction.cpp: Added.
* Source/WebCore/style/values/filter-effects/StyleBrightnessFunction.h: Added.
* Source/WebCore/style/values/filter-effects/StyleContrastFunction.cpp: Added.
* Source/WebCore/style/values/filter-effects/StyleContrastFunction.h: Added.
* Source/WebCore/style/values/filter-effects/StyleDropShadowFunction.cpp: Added.
* Source/WebCore/style/values/filter-effects/StyleDropShadowFunction.h: Added.
* Source/WebCore/style/values/filter-effects/StyleFilterProperty.cpp: Added.
* Source/WebCore/style/values/filter-effects/StyleFilterProperty.h: Added.
* Source/WebCore/style/values/filter-effects/StyleGrayscaleFunction.cpp: Added.
* Source/WebCore/style/values/filter-effects/StyleGrayscaleFunction.h: Added.
* Source/WebCore/style/values/filter-effects/StyleHueRotateFunction.cpp: Added.
* Source/WebCore/style/values/filter-effects/StyleHueRotateFunction.h: Added.
* Source/WebCore/style/values/filter-effects/StyleInvertFunction.cpp: Added.
* Source/WebCore/style/values/filter-effects/StyleInvertFunction.h: Added.
* Source/WebCore/style/values/filter-effects/StyleOpacityFunction.cpp: Added.
* Source/WebCore/style/values/filter-effects/StyleOpacityFunction.h: Added.
* Source/WebCore/style/values/filter-effects/StyleSaturateFunction.cpp: Added.
* Source/WebCore/style/values/filter-effects/StyleSaturateFunction.h: Added.
* Source/WebCore/style/values/filter-effects/StyleSepiaFunction.cpp: Added.
* Source/WebCore/style/values/filter-effects/StyleSepiaFunction.h: Added.
* Source/WebCore/style/values/primitives/StyleNone.cpp: Added.
* Source/WebCore/style/values/primitives/StyleNone.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Blending.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.cpp:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Logging.cpp: Added
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Logging.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes.h:
* Source/WebCore/style/values/text-decoration/StyleTextShadow.cpp: Added.
* Source/WebCore/style/values/text-decoration/StyleTextShadow.h: Added

Canonical link: <a href="https://commits.webkit.org/287509@main">https://commits.webkit.org/287509@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/c7c277ac2cd3cf3dc4d87c405c2252058faa2da5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79942 "40 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58939 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33339 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84456 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30916 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82050 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68003 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7235 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62478 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20305 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83009 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52537 "Found 1 new test failure: media/audioSession/getUserMedia.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72803 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42788 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49879 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26953 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29378 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/72916 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70999 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27433 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85888 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/79001 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7158 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5026 "Found 1 new test failure: pageoverlay/overlay-small-frame-mouse-events.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70750 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7335 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68643 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69995 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17432 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13991 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12925 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/101407 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7122 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12656 "Found 7 new failures in css/DeprecatedCSSOMFilterFunctionValue.h, css/CSSFilterPropertyValue.h, css/CSSTextShadowPropertyValue.h, css/CSSAppleColorFilterPropertyValue.h, css/DeprecatedCSSOMTextShadowValue.h, css/DeprecatedCSSOMBoxShadowValue.h, css/CSSBoxShadowPropertyValue.h and found 3 fixed files: css/CSSShadowValue.cpp, style/FilterOperationsBuilder.cpp, css/CSSShadowValue.h") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24727 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6969 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10481 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8773 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/css/parser/CSSPropertyParserConsumer+MetaConsumerDefinitions.h [hunk 1]
```diff
@@ -119,6 +119,20 @@ template<typename Raw> bool isValidCanonicalValue(Raw raw)
...
+{
+    static_assert(raw.range.options != CSS::RangeOptions::Default);
+
+    if constexpr (raw.range.options == CSS::RangeOptions::ClampLower) {
+        return { std::max<typename Raw::ValueType>(raw.value, raw.range.min) };
+    } else if constexpr (raw.range.options == CSS::RangeOptions::ClampUpper) {
+        return { std::min<typename Raw::ValueType>(raw.value, raw.range.max) };
+    } else if constexpr (raw.range.options == CSS::RangeOptions::ClampBoth) {
+        return { std::clamp<typename Raw::ValueType>(raw.value, raw.range.min, raw.range.max) };
+    }
```
Comment (darinadler) [inline Source/WebCore/css/parser/CSSPropertyParserConsumer+MetaConsumerDefinitions.h]: Not sure why these if statements have braces

File: Source/WebCore/css/parser/CSSPropertyParserConsumer+TextDecoration.cpp [hunk 2]
```diff
@@ -32,70 +32,113 @@
...
+
+    auto consumeLengths = [&] -> bool {
+        if (x)
+            return false;
+        x = MetaConsumer<CSS::Length<>>::consume(range, context, { }, lengthOptions);
+        if (!x)
+            return false;
+        y = MetaConsumer<CSS::Length<>>::consume(range, context, { }, lengthOptions);
+        if (!y)
+            return false;
```
Comment (darinadler) [inline Source/WebCore/css/parser/CSSPropertyParserConsumer+TextDecoration.cpp]: It's OK here that we consumed `x` as a side effect even though we returned `false`? I find these functions easier to understand when they either succeed or do nothing but I guess there are many variations.

File: Source/WebCore/css/values/CSSValueAggregates.h [hunk 3]
```diff
@@ -123,6 +134,11 @@ template<typename T, size_t inlineCapacity = 0> struct SpaceSeparatedVector {
     using const_reverse_iterator = typename Vector::const_reverse_iterator;
     using value_type = typename Vector::value_type;
 
+    SpaceSeparatedVector()
+        : value { }
+    {
+    }
```
Comment (darinadler) [inline Source/WebCore/css/values/CSSValueAggregates.h]: Is this different from `= default`?

File: Source/WebCore/css/values/CSSValueAggregates.h [hunk 4]
```diff
@@ -159,6 +175,11 @@ template<typename T, size_t inlineCapacity = 0> struct CommaSeparatedVector {
     using const_reverse_iterator = typename Vector::const_reverse_iterator;
     using value_type = typename Vector::value_type;
 
+    CommaSeparatedVector()
+        : value { }
+    {
+    }
```
Comment (darinadler) [inline Source/WebCore/css/values/CSSValueAggregates.h]: Is this different from `= default`?

File: Source/WebCore/style/values/StyleValueTypes.h [hunk 5]
```diff
@@ -511,6 +607,16 @@ template<typename StyleType, size_t inlineCapacity> struct Blending<SpaceSeparat
         }
         return true;
     }
+    auto canBlend(const SpaceSeparatedVector<StyleType, inlineCapacity>& a, const SpaceSeparatedVector<StyleType, inlineCapacity>& b, const RenderStyle& aStyle, const RenderStyle& bStyle) -> bool
+    {
+        if (a.size() != b.size())
+            return false;
+        for (size_t i = 0; i < a.size(); ++i) {
```
Comment (darinadler) [inline Source/WebCore/style/values/StyleValueTypes.h:614]: Zipped range?

File: Source/WebCore/style/values/StyleValueTypes.h [hunk 6]
```diff
@@ -520,6 +626,15 @@ template<typename StyleType, size_t inlineCapacity> struct Blending<SpaceSeparat
             result.append(WebCore::Style::blend(a[i], b[i], context));
         return { WTFMove(result) };
     }
+    auto blend(const SpaceSeparatedVector<StyleType, inlineCapacity>& a, const SpaceSeparatedVector<StyleType, inlineCapacity>& b, const RenderStyle& aStyle, const RenderStyle& bStyle, const BlendingContext& context) -> SpaceSeparatedVector<StyleType, inlineCapacity>
+    {
+        auto size = a.size();
+        typename SpaceSeparatedVector<StyleType, inlineCapacity>::Vector result;
+        result.reserveInitialCapacity(size);
+        for (size_t i = 0; i < size; ++i)
```
Comment (darinadler) [inline Source/WebCore/style/values/StyleValueTypes.h:634]: Zipper range?

File: Source/WebCore/style/values/StyleValueTypes.h [hunk 7]
```diff
@@ -534,6 +649,16 @@ template<typename StyleType, size_t inlineCapacity> struct Blending<CommaSeparat
         }
         return true;
     }
+    auto canBlend(const CommaSeparatedVector<StyleType, inlineCapacity>& a, const CommaSeparatedVector<StyleType, inlineCapacity>& b, const RenderStyle& aStyle, const RenderStyle& bStyle) -> bool
+    {
+        if (a.size() != b.size())
+            return false;
+        for (size_t i = 0; i < a.size(); ++i) {
```
Comment (darinadler) [inline Source/WebCore/style/values/StyleValueTypes.h:656]: Zipped range?

File: Source/WebCore/style/values/StyleValueTypes.h [hunk 8]
```diff
@@ -543,6 +668,15 @@ template<typename StyleType, size_t inlineCapacity> struct Blending<CommaSeparat
             result.append(WebCore::Style::blend(a[i], b[i], context));
         return { WTFMove(result) };
     }
+    auto blend(const CommaSeparatedVector<StyleType, inlineCapacity>& a, const CommaSeparatedVector<StyleType, inlineCapacity>& b, const RenderStyle& aStyle, const RenderStyle& bStyle, const BlendingContext& context) -> CommaSeparatedVector<StyleType, inlineCapacity>
+    {
+        auto size = a.size();
+        typename CommaSeparatedVector<StyleType, inlineCapacity>::Vector result;
+        result.reserveInitialCapacity(size);
+        for (size_t i = 0; i < size; ++i)
```
Comment (darinadler) [inline Source/WebCore/style/values/StyleValueTypes.h:676]: Zipped range?

**Review by darinadler — APPROVED** [review]
The proliferation of `CSSValue` classes concerns me. Is there a future where they go away?

File: Source/WebCore/css/parser/CSSPropertyParserConsumer+MetaConsumerDefinitions.h (hunk 1 above)
Comment (weinig) [reply Source/WebCore/css/parser/CSSPropertyParserConsumer+MetaConsumerDefinitions.h]: Good point!.

**Comment by weinig** [comment]
> The proliferation of `CSSValue` classes concerns me. Is there a future where they go away?

Yes. Once we get all the properties representable with a `struct` representation, the whole CSSValue hierarchy will go away (or at least the is the plan in my head).

I can reduce the number added here by 3 though, by making CSSFilterFunctionValue, CSSBoxShadowValue and CSSTextShadowValue instead be subclasses of DeprecatedCSSOMValue, since that it is their only reason for existing. I will do that.

File: Source/WebCore/css/values/CSSValueAggregates.h (hunk 4 above)
Comment (weinig) [reply Source/WebCore/css/values/CSSValueAggregates.h]: I don't think so. Will change.

File: Source/WebCore/css/values/CSSValueAggregates.h (hunk 3 above)
Comment (weinig) [reply Source/WebCore/css/values/CSSValueAggregates.h]: I don't think so. Will change.

File: Source/WebCore/css/parser/CSSPropertyParserConsumer+TextDecoration.cpp (hunk 2 above)
Comment (weinig) [reply Source/WebCore/css/parser/CSSPropertyParserConsumer+TextDecoration.cpp]: In this case, the only caller is doing a save/restore of the range based on the whole function, so it is ok, but I agree that making each function self sufficient in this regard is much better and makes refactoring more straightforward. I will fix this.

File: Source/WebCore/style/values/StyleValueTypes.h (hunk 5 above)
Comment (weinig) [reply Source/WebCore/style/values/StyleValueTypes.h:614]: Do we have a usable zippered range? I think the c++ ranges one is unavailable to us for now due to some macOS and iOS builds still using a relatively old libc++ (which didn't implement all the ranges features until later).

Comment (darinadler) [reply Source/WebCore/style/values/StyleValueTypes.h:614]: There's a file `ZippedRange.h`. The word _zippered_ lost some letters, it seems.

**Pull request merged**